File: 2020/02-password-philosophy/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

func solve(f string) (p1, p2 int) {
	input := read(f)
	for _, in := range input {
		cnt := strings.Count(in.password, in.char)
		if cnt >= in.in[0] && cnt <= in.in[1] {
			p1++
		}
	}

	for _, in := range input {
		if (in.password[in.in[0]-1] == in.char[0] && in.password[in.in[1]-1] != in.char[0]) ||
			(in.password[in.in[0]-1] != in.char[0] && in.password[in.in[1]-1] == in.char[0]) {
			p2++
		}
	}

	return p1, p2
}

type policy struct {
	password string
	char     string
	in       [2]int
}

func read(filename string) []policy {
	fc, _ := os.ReadFile(filename)

	lines := strings.Split(strings.TrimSpace(string(fc)), "\n")
	pa := []policy{}
	for _, line := range lines {
		pt := strings.Split(line, " ")
		p := policy{}
		json.Unmarshal([]byte("["+strings.Replace(pt[0], "-", ",", 1)+"]"), &p.in)
		p.char = string(pt[1][0])
		p.password = pt[2]

		pa = append(pa, p)
	}
	return pa
}

func main() {
	start := time.Now()
	fmt.Println("Day 02: Password Philosophy")
	p1, p2 := solve("input.txt")
	fmt.Println("\tPart One:", p1) // 398
	fmt.Println("\tPart Two:", p2) // 562
	fmt.Printf("Done in %.3f seconds \n", time.Since(start).Seconds())
}

File: 2020/02-password-philosophy/main.py
import time

def solve(file="input.txt"):
	p1, p2 = 0, 0
	lines = open(file, "r").read().splitlines()
	for i in range(len(lines)):
		parts = lines[i].split(":")
		pwd = parts[1].strip(" ")
		min, max = parts[0].split(" ")[0].split("-", 2)
		# Part 1:
		if  int(min) <= pwd.count(parts[0][-1]) <= int(max):
			p1+=1
		# Part 2:
		if (pwd[int(min)-1] == parts[0][-1]) ^ (pwd[int(max)-1] == parts[0][-1]): # XOR
			p2+=1

	return p1, p2


print("Day 02: Password Philosophy")
start = time.time()
p1, p2 = solve()
print("\tPart 1: ", p1)
print("\tPart 2: ", p2)
print("Done in %s seconds" % round(time.time() - start, 3))


